### ISSUE TITLE
Support Plugin detection for multiple locations (Xbox Game Pass)

### ISSUE THREAD

**Issue by Pickysaurus** [loot/loot#1861]
Related to: https://github.com/Nexus-Mods/NexusMods.App/issues/268

The LOOT app would benefit from some method of detecting plugins from multiple locations.

The ticket above described the situation with the Xbox Game Pass versions for FO4 and Skyrim SE where additional DLCs are in different folders but always loaded by the engine.

**Comment by Ortham** [comment]
Supporting this would be a PITA. From a quick scan of the code, LOOT assumes that plugins are in the game's data directory when:

- Checking if required files (including other plugins) are installed
- Reading extra Bash Tags supplied by a Bash Tag file accompanying the plugin
- Redating plugins
- Finding installed plugins

In libloot, it's assumed when:

- Checking if a plugin is valid
- Searching for archive files (BSA/BA2)
- Skipping loading the game's main master

In libloadorder, it's assumed when loading plugins as part of reading and writing the load order.

In loot-condition-interpreter, it's assumed when evaluating metadata conditions.

There are also security-adjacent concerns: LOOT currently prevents conditions from looking outside of the game's install path, basically to mitigate the possibility of anyone being persuaded to use malicious metadata that used information about their wider system to craft personalised messages (e.g. I see you're running an old version of software X, let me direct you to this perfectly trustworthy update). That restriction would have to be loosened to also allow reading in one of these paths (perhaps transparently?).

@Pickysaurus how does this impact the use of `.ccc` files to load the Creation Club plugins that are installed to these different folders? Also, do you know if other files present in the DLC install directories get detected, i.e. is it a kind of layering of the DLC directory on top of the game directory, or is it just that the game knows to look for these specific plugins in the DLC directory too?

**Comment by Pickysaurus** [comment]
The best I can tell, the `.ccc` file is not in use for the Xbox version. These are the directory listings for the game and the DLC

### Skyrim SE game directory
```
Data/
Skyrim/
appxmanifest.xml
atimgpud.dll
bink2w64.dll
binkw32.dll
binkw64.dll
d3dcompiler_46.dll
d3dcompiler_47.dll
DejaInsight.x64.dll
gamelaunchhelper.exe
High.ini
layout_dd5e0e3d-76e9-4b68-b133-68fa14bd9601.xml
Low.ini
MainTitle.wav
Medium.ini
MicrosoftGame.Config
p4com64.dll
Skyrim.ccc
SkyrimSE.exe
SkyrimSELauncher.exe
SkyrimSELauncher.exe.manifest
Skyrim_Default_cn.ini
Skyrim_Default_de.ini
Skyrim_Default_en.ini
Skyrim_Default_es.ini
Skyrim_Default_fr.ini
Skyrim_Default_it.ini
Skyrim_Default_ja.ini
Skyrim_Default_pl.ini
Skyrim_Default_ru.ini
SSE_logo.png
SSE_smalllogo.png
SSE_splashscreen.png
SSE_storelogo.png
Ultra.ini
XCurl.dll
```

### DLC directory
```
Images/
appxmanifest.xml
ccAARSSE001-Manufactory.esl
ccASVSSE001-ALMSIVI.esl
ccASVSSE001-ALMSIVI.esm
ccBGSSSE002-ExoticArrows.esl
ccBGSSSE003-Zombies.esl
ccBGSSSE004-RuinsEdge.esl
ccBGSSSE005-Goldbrand.esl
ccBGSSSE006-StendarsHammer.esl
ccBGSSSE007-Chrysamere.esl
ccBGSSSE008-Wraithguard.esl
ccBGSSSE010-PetDwarvenArmoredMudcrab.esl
ccBGSSSE011-HrsArmrElvn.esl
ccBGSSSE012-HrsArmrStl.esl
ccBGSSSE013-Dawnfang.esl
ccBGSSSE014-SpellPack01.esl
ccBGSSSE016-Umbra.esm
ccBGSSSE017-SwordofJyggalag.esl
ccBGSSSE018-Shadowrend.esl
ccBGSSSE019-StaffofSheogorath.esl
ccBGSSSE020-GrayCowl.esl
ccBGSSSE021-LordsMail.esl
ccBGSSSE025-AdvDSGS.esm
ccBGSSSE029-Nerveshatter.esl
ccBGSSSE031-AdvCyrus.esm
ccBGSSSE032-MntDaed.esl
ccBGSSSE034-MntUni.esl
ccBGSSSE035-PetNHound.esl
ccBGSSSE036-PetBWolf.esl
ccBGSSSE037-Curios.esl
ccBGSSSE040-AdvObGobs.esl
ccBGSSSE041-NetchLeather.esl
ccBGSSSE043-CrossElv.esl
ccBGSSSE045-Hasedoki.esl
ccBGSSSE050-BA_Daedric.esl
ccBGSSSE051-BA_DaedricMail.esl
ccBGSSSE052-BA_Iron.esl
ccBGSSSE053-BA_Leather.esl
ccBGSSSE054-BA_Orcish.esl
ccBGSSSE055-BA_OrcishScaled.esl
ccBGSSSE056-BA_Silver.esl
ccBGSSSE057-BA_Stalhrim.esl
ccBGSSSE058-BA_Steel.esl
ccBGSSSE059-BA_Dragonplate.esl
ccBGSSSE060-BA_Dragonscale.esl
ccBGSSSE061-BA_Dwarven.esl
ccBGSSSE062-BA_DwarvenMail.esl
ccBGSSSE063-BA_Ebony.esl
ccBGSSSE064-BA_Elven.esl
ccCBHSSE001-Gaunt.esl
ccEDHSSE001-NorJewel.esl
ccEDHSSE002-SplKntSet.esl
ccEEJSSE001-Hstead.esm
ccEEJSSE002-Tower.esl
ccEEJSSE003-Hollow.esl
ccEEJSSE004-Hall.esl
ccFFBSSE001-ImperialDragon.esl
ccFFBSSE002-CrossbowPack.esl
ccFFBSSE002-StormBear.esl
ccFSVSSE001-Backpacks.esl
ccKRTSSE001_Altar.esl
ccMDRSSE001-DweSanctuary.esm
ccMTYSSE001-KnightsOfTheNine.esl
ccMTYSSE002-VE.esl
ccPEWSSE002-ArmsOfChaos.esl
ccQDRSSE001-SurvivalMode.esl
ccQDRSSE002-Firewood.esl
ccRMSSSE001-NecroHouse.esl
ccTWBSSE001-PuzzleDungeon.esm
ccVSVSSE001-Winter.esl
ccVSVSSE002-Pets.esl
ccVSVSSE003-NecroArts.esl
gamelaunchhelper.exe
layout_903d1de8-330a-0ea8-e344-823ebb4c3b17.xml
MicrosoftGame.config
```

I'll test out the layering theory, however, I refunded the Anniversary Upgrade for Skyrim so I'll have to redownload Fallout 4 and its DLCs to test. 

I appreciate this isn't an easy change, but with Starfield and TES6 coming to the Xbox app too, this is very likely to be a pattern that persists going forward. I predict that when these games come out, Xbox players will be the second most common after Steam.

**Comment by Ortham** [comment]
> I appreciate this isn't an easy change, but with Starfield and TES6 coming to the Xbox app too, this is very likely to be a pattern that persists going forward. I predict that when these games come out, Xbox players will be the second most common after Steam.

Unfortunately I think you're right. Thanks for the info, it'll be useful.

**Comment by Pickysaurus** [comment]
Thanks, 

I just tested the layering theory and it seems that only the specific files for the DLC are layered over the base game.

I created test.esp in xEdit, then dropped it into the Wasteland Workshop data folder. 

On loading the game and opening the "Mods" -> "Load order" screen, no plugins were shown.

On opening the DLCs screen, Wasteland Workshop was recognised.

**Comment by Ortham** [comment]
I've bought Skyrim SE and the Anniversary Upgrade DLC on the Microsoft Store (it's more expensive there than on other stores...), and the files are split into different folders the same as described when playing Skyrim SE through Xbox Game Pass (not that I expected how you pay for it to matter).

**Comment by sibir-ine** [comment]
> The base The Elder Scrolls V- Skyrim Special Edition (PC) path contains a few CC plugins: 

These are the four free CC plugins that came with the Anniversary Update (i.e., v1.6.x) rather than the Anniversary Edition upgrade.

Here's a list of files in my Data folder from a fresh install of the AE upgrade from Steam with their CRC32 checksums & file sizes:

<details>
<summary>Steam AE file list</summary>

CRC32 | Size (bytes) | Name
----- | ------------ | ----
AA6A9CF0 | 4758160 | ccafdsse001-dwesanctuary.bsa
5442BD0C | 523532 | ccafdsse001-dwesanctuary.esm
15523886 | 524046246 | ccasvsse001-almsivi.bsa
D45403A6 | 649488 | ccasvsse001-almsivi.esm
BFF60BFD | 377673669 | ccbgssse001-fish.bsa
3F2DC672 | 1425350 | ccbgssse001-fish.esm
BBB4622B | 69803784 | ccbgssse002-exoticarrows.bsa
51D308C0 | 182186 | ccbgssse002-exoticarrows.esl
3A2A2DC4 | 112488634 | ccbgssse003-zombies.bsa
FE6BE367 | 180899 | ccbgssse003-zombies.esl
5BFCF5EE | 40101322 | ccbgssse004-ruinsedge.bsa
5EE1F7FA | 7913 | ccbgssse004-ruinsedge.esl
5459FFDA | 11231646 | ccbgssse005-goldbrand.bsa
832A954E | 213662 | ccbgssse005-goldbrand.esl
99367781 | 33736697 | ccbgssse006-stendarshammer.bsa
B2139281 | 4539 | ccbgssse006-stendarshammer.esl
F94A2402 | 4893203 | ccbgssse007-chrysamere.bsa
E4DB0C51 | 54185 | ccbgssse007-chrysamere.esl
92458B19 | 37750618 | ccbgssse008-wraithguard.bsa
2B01BC19 | 240477 | ccbgssse008-wraithguard.esl
4C4CC364 | 45198131 | ccbgssse010-petdwarvenarmoredmudcrab.bsa
00A63157 | 10489 | ccbgssse010-petdwarvenarmoredmudcrab.esl
5431051B | 15948258 | ccbgssse011-hrsarmrelvn.bsa
4D90586A | 896 | ccbgssse011-hrsarmrelvn.esl
750BC740 | 12638944 | ccbgssse012-hrsarmrstl.bsa
4287AA21 | 910 | ccbgssse012-hrsarmrstl.esl
0DDE8C62 | 28625104 | ccbgssse013-dawnfang.bsa
C4E35D25 | 244256 | ccbgssse013-dawnfang.esl
52258038 | 22249957 | ccbgssse014-spellpack01.bsa
612F5721 | 45414 | ccbgssse014-spellpack01.esl
430AC2CE | 10284176 | ccbgssse016-umbra.bsa
A9D2211B | 471115 | ccbgssse016-umbra.esm
2EFDCFF0 | 73376503 | ccbgssse018-shadowrend.bsa
2256DD07 | 253552 | ccbgssse018-shadowrend.esl
0C1FC950 | 11887694 | ccbgssse019-staffofsheogorath.bsa
FCBD8672 | 172972 | ccbgssse019-staffofsheogorath.esl
13B14F43 | 15063067 | ccbgssse020-graycowl.bsa
5231E7EF | 70659 | ccbgssse020-graycowl.esl
F3F806F1 | 45873038 | ccbgssse021-lordsmail.bsa
3755436F | 113076 | ccbgssse021-lordsmail.esl
A52E6E81 | 1092876206 | ccbgssse025-advdsgs.bsa
A3D900E1 | 812763 | ccbgssse025-advdsgs.esm
A8A7376E | 145481727 | ccbgssse031-advcyrus.bsa
BF898B88 | 855137 | ccbgssse031-advcyrus.esm
66345C03 | 76553696 | ccbgssse034-mntuni.bsa
39037200 | 146032 | ccbgssse034-mntuni.esl
94152A9C | 41641998 | ccbgssse035-petnhound.bsa
DA4CBDA5 | 201862 | ccbgssse035-petnhound.esl
4DE069D2 | 17949273 | ccbgssse036-petbwolf.bsa
DF613F47 | 135117 | ccbgssse036-petbwolf.esl
84393161 | 111740475 | ccbgssse037-curios.bsa
EA917D0B | 37476 | ccbgssse037-curios.esl
B3131A8C | 7056850 | ccbgssse038-bowofshadows.bsa
CF70E2A1 | 114743 | ccbgssse038-bowofshadows.esl
2B654C6F | 77425181 | ccbgssse040-advobgobs.bsa
86A233E1 | 316111 | ccbgssse040-advobgobs.esl
2D36281D | 122116053 | ccbgssse041-netchleather.bsa
87F09B67 | 263377 | ccbgssse041-netchleather.esl
C804C342 | 34052447 | ccbgssse043-crosselv.bsa
9F926A70 | 143881 | ccbgssse043-crosselv.esl
3A156EA9 | 6471111 | ccbgssse045-hasedoki.bsa
E75D726D | 117019 | ccbgssse045-hasedoki.esl
751A6EA0 | 26519981 | ccbgssse050-ba_daedric.bsa
FA1BF723 | 118387 | ccbgssse050-ba_daedric.esl
A9986D4D | 47651851 | ccbgssse051-ba_daedricmail.bsa
ED0DAC61 | 121476 | ccbgssse051-ba_daedricmail.esl
52A6F765 | 41064189 | ccbgssse052-ba_iron.bsa
44CC9684 | 114892 | ccbgssse052-ba_iron.esl
6786300C | 35802145 | ccbgssse053-ba_leather.bsa
C280B67C | 115735 | ccbgssse053-ba_leather.esl
977EAB82 | 43448086 | ccbgssse054-ba_orcish.bsa
1799B889 | 25666 | ccbgssse054-ba_orcish.esl
40279276 | 43781890 | ccbgssse055-ba_orcishscaled.bsa
19E55E2C | 29188 | ccbgssse055-ba_orcishscaled.esl
569CF9F2 | 48153667 | ccbgssse056-ba_silver.bsa
A67387E9 | 114511 | ccbgssse056-ba_silver.esl
64C2DD93 | 29322687 | ccbgssse057-ba_stalhrim.bsa
B0337665 | 176136 | ccbgssse057-ba_stalhrim.esl
1B999EAC | 50766738 | ccbgssse058-ba_steel.bsa
482C80EF | 120621 | ccbgssse058-ba_steel.esl
1091C40D | 55387235 | ccbgssse059-ba_dragonplate.bsa
A6E40E02 | 131189 | ccbgssse059-ba_dragonplate.esl
0DFBA9A8 | 40386574 | ccbgssse060-ba_dragonscale.bsa
FFCEF134 | 116459 | ccbgssse060-ba_dragonscale.esl
27162549 | 39409700 | ccbgssse061-ba_dwarven.bsa
96DA6A08 | 111160 | ccbgssse061-ba_dwarven.esl
4200B2CD | 32992326 | ccbgssse062-ba_dwarvenmail.bsa
C7B82470 | 114053 | ccbgssse062-ba_dwarvenmail.esl
C79B9337 | 47723042 | ccbgssse063-ba_ebony.bsa
28D1727F | 134877 | ccbgssse063-ba_ebony.esl
EC284CA2 | 45841470 | ccbgssse064-ba_elven.bsa
F67CFF9B | 111663 | ccbgssse064-ba_elven.esl
DAC6D7BB | 47139906 | ccbgssse066-staves.bsa
F19EB095 | 272311 | ccbgssse066-staves.esl
607BAB70 | 386611786 | ccbgssse067-daedinv.bsa
4D769AAB | 1184014 | ccbgssse067-daedinv.esm
568B4F3B | 11540866 | ccbgssse068-bloodfall.bsa
767BDEC9 | 12087 | ccbgssse068-bloodfall.esl
E5FCC808 | 42851238 | ccbgssse069-contest.bsa
D2910DFE | 17988 | ccbgssse069-contest.esl
2D14795A | 68166604 | cccbhsse001-gaunt.bsa
B2682AA4 | 28801 | cccbhsse001-gaunt.esl
70D7F078 | 9712736 | ccedhsse001-norjewel.bsa
1059E62C | 196569 | ccedhsse001-norjewel.esl
8DBDF13E | 89101366 | ccedhsse002-splkntset.bsa
FCF374B9 | 160351 | ccedhsse002-splkntset.esl
F034AE98 | 86296244 | ccedhsse003-redguard.bsa
802755BA | 151347 | ccedhsse003-redguard.esl
BCC3B083 | 28511474 | cceejsse001-hstead.bsa
8D907DED | 354470 | cceejsse001-hstead.esm
696BD13B | 66127028 | cceejsse002-tower.bsa
7D7DE5B5 | 399976 | cceejsse002-tower.esl
BE763B17 | 28529933 | cceejsse003-hollow.bsa
A7C5D720 | 309934 | cceejsse003-hollow.esl
358CBDD0 | 59590753 | cceejsse004-hall.bsa
FBD7033E | 382824 | cceejsse004-hall.esl
B4EE0425 | 29777150 | cceejsse005-cave.bsa
2A44A95B | 422522 | cceejsse005-cave.esm
1FE190AF | 177912120 | ccffbsse001-imperialdragon.bsa
B61EB1F5 | 207155 | ccffbsse001-imperialdragon.esl
2B968202 | 91832685 | ccffbsse002-crossbowpack.bsa
F3442D86 | 22307 | ccffbsse002-crossbowpack.esl
B5AF2602 | 57216555 | ccfsvsse001-backpacks.bsa
515E986E | 18898 | ccfsvsse001-backpacks.esl
6F1A7B14 | 132317834 | cckrtsse001_altar.bsa
CE4A252F | 499020 | cckrtsse001_altar.esl
E1AB8832 | 80681749 | ccmtysse001-knightsofthenine.bsa
3D17B0EB | 87925 | ccmtysse001-knightsofthenine.esl
434F5158 | 118025497 | ccmtysse002-ve.bsa
90AD3545 | 259928 | ccmtysse002-ve.esl
7B9A8ADD | 31141295 | ccpewsse002-armsofchaos.bsa
D8F7F985 | 141394 | ccpewsse002-armsofchaos.esl
2EC781F8 | 12835601 | ccqdrsse001-survivalmode.bsa
87023079 | 240724 | ccqdrsse001-survivalmode.esl
596F0B8C | 443268 | ccqdrsse002-firewood.bsa
31D3A629 | 264689 | ccqdrsse002-firewood.esl
7F95704B | 51200165 | ccrmssse001-necrohouse.bsa
BAFF0369 | 393319 | ccrmssse001-necrohouse.esl
6BC4CBFB | 362189431 | cctwbsse001-puzzledungeon.bsa
25D6FE7B | 1411214 | cctwbsse001-puzzledungeon.esm
6764C825 | 42538666 | ccvsvsse001-winter.bsa
DEC857BE | 115940 | ccvsvsse001-winter.esl
B430C8DF | 102744267 | ccvsvsse002-pets.bsa
474A4A90 | 186345 | ccvsvsse002-pets.esl
FB5CA910 | 169853815 | ccvsvsse003-necroarts.bsa
8D0AE774 | 224610 | ccvsvsse003-necroarts.esl
5AAF39EA | 18261078 | ccvsvsse004-beafarmer.bsa
C84E56C2 | 485046 | ccvsvsse004-beafarmer.esl
641044FE | 25884627 | Dawnguard.esm
A959AA60 | 64663944 | Dragonborn.esm
A9910F5C | 3977660 | HearthFires.esm
5669A8E3 | 65101236 | Skyrim - Animations.bsa
1A329BDB | 105751175 | Skyrim - Interface.bsa
BA40ABF9 | 1153090124 | Skyrim - Meshes0.bsa
4BEBA3AB | 378853391 | Skyrim - Meshes1.bsa
489EE525 | 204987429 | Skyrim - Misc.bsa
94CC86F7 | 67308970 | Skyrim - Shaders.bsa
C6D453EB | 1538656059 | Skyrim - Sounds.bsa
0731E140 | 652401558 | Skyrim - Textures0.bsa
A6D107C4 | 1511504224 | Skyrim - Textures1.bsa
543F9DE5 | 1346030841 | Skyrim - Textures2.bsa
465FC3C4 | 1401677791 | Skyrim - Textures3.bsa
7FE69495 | 1265325033 | Skyrim - Textures4.bsa
948BB0EF | 809121183 | Skyrim - Textures5.bsa
47054551 | 107211187 | Skyrim - Textures6.bsa
4674A351 | 700733354 | Skyrim - Textures7.bsa
758FC236 | 214496000 | Skyrim - Textures8.bsa
D882EE28 | 1807969854 | Skyrim - Voices_en0.bsa
66212483 | 249753351 | Skyrim.esm
0F9F1100 | 18857707 | Update.esm
7D220AE7 | 13835808 | Video\BGS_Logo.bik

</details>

**Comment by Ortham** [comment]
> These are the four free CC plugins that came with the Anniversary Update (i.e., v1.6.x) rather than the Anniversary Edition upgrade.

That makes sense, and thanks for the details. 

It looks like the Steam AE DLC has the same CC files as my GOG install - I checked a few CRCs were equal, but didn't check most of them. I did check all the CRCs of the base Skyrim BSAs, and a couple of them in are different in my GOG install:

| CRC | Size / bytes | Filename |
|-----|-------------|----------|
| 81DFB40C | 105392692 | Skyrim - Interface.bsa
| ADF7A4DE | 205029198 | Skyrim - Misc.bsa

And in my MS Store install they are:

| CRC | Size / bytes | Filename |
|-----|-------------|----------|
| EAAFB29A | 105591873 | Skyrim - Interface.bsa
| 59B79013 | 204987429 | Skyrim - Misc.bsa

I was half-expecting the MS Store BSAs to match the Steam BSAs as both have access to the Creation Club and Steam Workshop mods (not sure how that works for the MS Store version...), while the GOG version does not, but that's apparently not it.

I also had a look at the checksums of the Anniversary Update plugins in my different installs:

| File | Steam CRC | GOG CRC | MS Store (SE) | MS Store (AE DLC) |
|-----|------------|------------|--------------|---------------------|
| ccBGSSSE001-Fish.bsa | BFF60BFD | BFF60BFD | BFF60BFD | N/A
| ccBGSSSE001-Fish.esm | 3F2DC672 | 3F2DC672 | 3F2DC672 | N/A
| ccBGSSSE025-AdvDSGS.bsa | A52E6E81 | A52E6E81 | A52E6E81 | N/A
| ccBGSSSE025-AdvDSGS.esm | A3D900E1 | A3D900E1 | A3D900E1 | 264006CB
| ccBGSSSE037-Curios.bsa | 84393161 | 84393161 | 84393161 | N/A
| ccBGSSSE037-Curios.esl | EA917D0B | EA917D0B | EA917D0B | 886B69D1
| ccQDRSSE001-SurvivalMode.bsa | 2EC781F8 | 2EC781F8 | 2EC781F8 | N/A
| ccQDRSSE001-SurvivalMode.esl | 87023079 | 87023079 | 87023079 | 611D6AA2

One other thing I noticed is that the MS Store's `Skyrim.ccc` file includes CC plugins that are not included in the AE DLC - presumably this is because I guess they're still available to buy through Bethesda's in-game store though.

I still haven't found where the BSAs for the AE DLC are. I used Process Monitor to see what files Skyrim was trying to load when it ran, and weirdly `svchost.exe` and `GamingServices.exe` both accessed the `Skryim 10 - Creation Club Full Content` directory, but nothing tried loading the ESL files in it, or any BSAs. `SkyrimSE.exe` did try to load all the plugins listed in `Skyrim.ccc`. So now I'm doubting if the AE DLC is installed properly or usable by the game at all - I did verify my install in the Xbox app, but that didn't change anything.

**Comment by sibir-ine** [comment]
> both have access to the Creation Club and Steam Workshop mods

There's no Steam Workshop for SE, only LE. You might be thinking of Bethesda's modding platform.

**Comment by Ortham** [comment]
OK, after more investigating I found that I can go into the Creation Club section in-game and see that all the AE CC content is listed as "purchased" there (including content that's not found in the `Skryim 10 - Creation Club Full Content` path but is included in Steam and GOG installs), and I can download them through that interface. Doing so downloads them into `The Elder Scrolls V- Skyrim Special Edition (PC)\Content\Data` putting them in the same directory as all the base game files. The newly download files have checksums that match those in my GOG install (and not those in the `Skryim 10 - Creation Club Full Content` path).

I'm pretty sure the "purchased" Creation Club content came from buying the AE DLC in the Microsoft Store, because I don't own AE on Steam (I did try launching Skyrim SE there to see if I now owned the AE CC content there too, but it's stuck on a spinner before the game launch reaches the title screen) and AE on GOG doesn't include the options to download content from Bethesda's platform.

So in the end I think the MS Store's creation of the `Skryim 10 - Creation Club Full Content` path and all the files in it is an unnecessary distraction, and the AE DLC that you can buy in the MS Store is actually about marking you as owning the CC content in the in-game storefront, allowing you to download it through the game.

As such, I don't think there's anything to be done here: there's no need to support plugins in multiple locations, because the second location is irrelevant in practice.

I've only been able to test that for Skyrim though, hopefully @Pickysaurus can check if the same is true for FO4.

**Comment by Pickysaurus** [comment]
Thanks for testing, I'm afraid this still 100% applies for Fallout 4's Game of the Year DLCs

**Comment by Pickysaurus** [comment]
> @Pickysaurus I know you said that the DLCs for FO4 were recognised in-game, but given the above, are you able to access their content in-game? As far as I can tell, the Skyrim AE DLC doesn't actually work.

I was able to `coc` into Far Harbour with the DLCs loaded from the Xbox store.

**Comment by Pickysaurus** [comment]
A good test might be to try `coc`-ing to the new dungeon added by one of the creations with the CC files in your game folder and without.

**Comment by Ortham** [comment]
> I was able to `coc` into Far Harbour with the DLCs loaded from the Xbox store.

I bought and downloaded FO4 GOTY and can see it works as you described (though I haven't tried adding other plugins to the DLC directories). 

As far as I can tell though, the AE DLC for Skyrim SE just plain doesn't work in that way: maybe it was intended and it's just broken, or maybe they started trying to go that route but switched to expecting people to download the content in-game but didn't clean up after themselves, I don't know. It doesn't really matter, since the changes I need to make for FO4 will be generic, so Skyrim SE not being affected really just means less config.

Hopefully I can figure out an approach and test against the games quickly so that I can get my ~ £63 refunded...

**Comment by Pickysaurus** [comment]
@Ortham Please go ahead and claim the refund. I'm happy to give you info or maybe even remote access to my system to test things if that helps? 

I'd definitely prefer you weren't out of pocket over this!

**Comment by Ortham** [comment]
I *think* I've made all the necessary changes:

- loot-condition-interpreter now takes a list of data paths to use, and when given a file path in a condition it'll treat the path as relative to one of the data paths, trying each of the data paths in the order they're given, stopping at the first that the file path exists in. The many condition looks across all the configured data paths, so if you've got one match in one data path and one in another, the condition will evaluate to true.
- libloadorder will now look for the Fallout 4 DLC plugins in their separate directories (resolved relative to the game install path) if the Fallout 4 install looks like a Microsoft Store install (i.e. if `appxmanifest.xml` is present in the game install path). This is done in two ways: 
  - if trying to load a plugin given a filename (e.g. read from `plugins.txt`), it's matched case-insensitively against the Fallout 4 DLC plugin filenames.
  - when scanning for plugins to load, libloadorder will first look for the DLC plugins in their separate directories before scanning the game's Data directory for plugins.
- libloadorder also supports overriding the recognised external plugin paths, so its support is not limited to the MS Store Fallout 4 DLC plugins (though specific plugin paths need to be given, you can't provide directories for it to scan other than the game's plugins directory).
- libloot now takes a vector of plugin paths when loading plugins and sorting them: relative paths are resolved relative to the game's data path, and absolute plugins are used as given, so it's backwards-compatible with the previous semantics, where they took a vector of plugin filenames. In the same way, checking if a given plugin is valid now also takes a plugin path instead of a filename.
- libloot passes the (hardcoded) Fallout 4 DLC data paths followed by the game data path to loot-condition-interpreter when the game is Fallout 4 and `appxmanifest.xml` is present, otherwise it just passes the game data path as before. The paths are passed in that order so that the DLC paths take precedence over the main data path.
- libloot also now checks that the plugins passed to load or sort have unique filenames, which has always been necessary but was much less likely to be a problem with all the plugins in one directory.
- libloot now has a hardcoded list of Fallout 4 DLC `.ba2` file paths (relative to the game install path) that it now loops through (checking if they exist) when scanning for BSA/BA2 files, if the game is Fallout 4 and `appxmanifest.xml` is present.
- LOOT copies libloadorder's behaviour when looking for installed plugins and when resolving the path of a given plugin filename.

It would be good if I could avoid duplicating the relative directory and file paths in LOOT, libloot and libloadorder, but that would mean worse default behaviour for the latter two. It would probably be useful in the future to have some way of overriding the hardcoded recognised DLC paths in libloot, but that would mean breaking API/ABI changes, so I'm holding off on that for now. I'd also like to convert some `std::string`s to `std::filesystem::path`s, but haven't for the same reason.

LOOT can find the game and its DLCs and sort them:

![image](https://user-images.githubusercontent.com/2430738/233454797-7e02ec16-9f1e-42a2-a2e4-5bbb1e6ec805.png)

I've still got a bunch of tests to write for LOOT and some cleanup to do though.

> I'm happy to give you info or maybe even remote access to my system to test things if that helps?

Thanks, I've got Skyrim SE refunded and I'll hold on to FO4 for a few more days, but after that I may bug you for more info and/or testing.

**Comment by Ortham** [comment]
I've been doing a bit more investigation into how FO4 loads plugins, and it looks like FO4 doesn't just look for the DLC files specifically, instead it searches the DLC install paths for files to load.

I moved the plugin and BA2 files for the Contraptions Workshop DLC to the Automatron DLC data path and was still able to load the game and add an item added by Contraptions to my inventory. I also added the Unofficial Fallout 4 Patch to the Nuka-World data path and was able to see its changes in-game, though it didn't appear in the in-game mods UI.

Process Monitor shows that FO4 looks for `Fallout4.esm`, CC plugins and BA2 files (from the base game, DLCs and the unofficial patch) in all the DLC data paths as well as the FO4 data path, in the following order, stopping when it finds the file:

```
D:\Games\XboxGames\Fallout 4- Automatron (PC)\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4- Nuka-World (PC)\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4- Wasteland Workshop (PC)\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4- High Resolution Texture Pack\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4- Vault-Tec Workshop (PC)\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4- Far Harbor (PC)\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4- Contraptions Workshop (PC)\Content\Data\Fallout4.esm
D:\Games\XboxGames\Fallout 4 (PC)\Content\Data\Fallout4.esm
```

Fallout4.exe also looks for `Fallout4.esm` and CC plugins in the Content directory (one up from Data) if they're not found in the Data directory. The same is true for loose resource files (e.g. `*Geometry.psg` and `*.cdx` files), but not DLC BA2 files (aside from `*Voices_en.ba2`).

Process Monitor also shows that `Fallout4.exe` queries the DLC Data directories for files and folders and finds whatever is in there, in this order:

```
D:\Games\XboxGames\Fallout 4- Contraptions Workshop (PC)\Content\Data\*
D:\Games\XboxGames\Fallout 4- Far Harbor (PC)\Content\Data\*
D:\Games\XboxGames\Fallout 4- Vault-Tec Workshop (PC)\Content\Data\*
D:\Games\XboxGames\Fallout 4- High Resolution Texture Pack\Content\Data\*
D:\Games\XboxGames\Fallout 4- Wasteland Workshop (PC)\Content\Data\*
D:\Games\XboxGames\Fallout 4- Nuka-World (PC)\Content\Data\*
D:\Games\XboxGames\Fallout 4- Automatron (PC)\Content\Data\*
```

There doesn't seem to be a `Data\*` query for the base game path.

---

libloadorder, libloot and LOOT assume that DLC plugins are installed to their default directory, but the above shows that isn't actually true, so I can't just match against a DLC plugin filename and know what path it's at, instead I need to check where it actually exists.

- libloadorder's approach should actually still work, because it's configured with a list of external plugin paths and filenames get matched against those paths' filenames, so it's up to the user of the library to supply the correct paths. It would be more useful if that didn't need to be configured and the default config supported more than just the vanilla DLC plugin paths, but that would mean having to loop through the external data paths checking if the file exists, just to get a path to then read, and that'll need to be done whilst still supporting ghosted plugins. It's probably still best to make that change though.
- libloot has a hardcoded list of data paths to pass down to loot-condition-interpreter, which is fine. It's also got a hardcoded list of DLC BA2 files, which is not fine and instead it should just scan all those hardcoded data paths for archive files. It also doesn't expose any way to provide external plugin paths (to avoid breaking ABI) and so can't support non-vanilla DLC install directories because it can't tell libloadorder about any non-vanilla plugins in them.
- LOOT has a hardcoded list of DLC plugin paths, which needs to be replaced by scanning those directories. It also resolves plugin paths by comparing against hardcoded DLC filenames, which won't work with non-vanilla DLC install directories, so that needs to be changed to check if the file exists in each data path until it's found.
